AWS Glue Crawler Integration

marshall7m opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
## Description

This PR integrates an AWS glue crawler operator, hook and sensor that can be used to trigger glue crawlers from Airflow.  

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

marshall7m committed:
initial glue crawler operator/hook

marshall7m committed:
add operator/hook tests and add working operator/hook

marshall7m committed:
add glue crawler sensor

marshall7m committed:
use lazy formatting for logs and rm trail whitespaces

marshall7m committed:
fix styling/formatting issues and add crawler sensor test

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+class AwsGlueCrawlerHook(AwsBaseHook):
+    """
+    Interacts with AWS Glue Crawler
+    :param crawler_name = Unique crawler name per AWS Account
```
```suggestion

    :param crawler_name: Unique crawler name per AWS Account
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+    """
+    Interacts with AWS Glue Crawler
+    :param crawler_name = Unique crawler name per AWS Account
+    :type crawler_name = Optional[str]
```
```suggestion
    :type crawler_name: Optional[str]
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+        """
+        Get state of the Glue crawler. The crawler state can be
+        ready, running, or stopping.
+        :param crawler_name: unique crawler name per AWS account
```
```suggestion

        :param crawler_name: unique crawler name per AWS account
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+        """
+        Get current status of the Glue crawler. The crawler
+        status can be succeeded, cancelled, or failed.
+        :param crawler_name: unique crawler name per AWS account
```
```suggestion

        :param crawler_name: unique crawler name per AWS account
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+        Waits until Glue crawler with crawler_name completes or
+        fails and returns final state if finished.
+        Raises AirflowException when the crawler failed
+        :param crawler_name: unique crawler name per AWS account
```
```suggestion

        :param crawler_name: unique crawler name per AWS account
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+    def get_or_create_glue_crawler(self) -> str:
+        """
+        Creates the crawler if the crawler doesn't exists and returns the crawler name
+        :return:Name of the crawler
```
```suggestion

        :return :Name of the crawler
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,282 @@
+    def get_crawler_metrics(self, crawler_name):
+        """
+        Prints crawl runtime and glue catalog table metrics associated with the crawler
+        :return:Dictionary of all the crawler metrics
```
```suggestion

        :return: Dictionary of all the crawler metrics
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,288 @@
+                    print('Tables Deleted: ', metrics['TablesDeleted'])
+                    
+                    return {'Status': crawler_run_status}
+                    
```
```suggestion

                    return {'Status': crawler_run_status}

```
Whitespace

marshall7m committed:
fix logging format and trailing whiespaces

marshall7m committed:
add update crawler feature

marshall7m committed:
add overwrite param docs to glue crawler hook

marshall7m committed:
add logging and error catching for updating crawler

marshall7m committed:
add glue crawler to list of amz providers

marshall7m committed:
consolidate operator/hook params into dict config

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,294 @@
+class AwsGlueCrawlerHook(AwsBaseHook):
+    """
+    Interacts with AWS Glue Crawler
+    :param crawler_name = Unique crawler name per AWS Account
```
generally i think hook init params are focused on authentication and connection behavior

i think it might make sense to move most of these kwargs to the `get_or_create_crawler` method

sort of like how with s3hook you don't provide key and bucket at init, or with sql hook you don't provide sql at init -- but you provide methods for interacting with the service.

then you could instantiate a hook once and use it to deal with multiple crawlers.  

then, re your slack question about too many params,  perhaps the method becomes `get_or_create_crawler(**crawler_kwargs)`

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,294 @@
+        Initializes connection with AWS Glue to run crawler
+        :return:
+        """
+        glue_client = self.get_conn()
```
one thing i like to do in hooks is create a cached propert such as `client` or `glue_client` or in other hooks `session`, so that you don't need to remember to call `get_conn` in every case, and you can just reuse the same authenticated object without reauthenticating.

e.g. 

```python
@cached_property
def glue_client(self):
    return self.get_conn()
```

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,294 @@
+        crawler_run_status = crawler_run['Crawler']['LastCrawl']['Status']
+        return crawler_run_status
+
+    def crawler_completion(self, crawler_name: str) -> str:
```
maybe `wait_for_crawler_completion` or `await_crawler_completion` or `await_completion` or `wait_for_completion` ---- something to indicated what it will do

dstandish commented on airflow/providers/amazon/aws/operators/glue_crawler.py:
```diff
@@ -0,0 +1,164 @@
+        glue_catalog_targets_configuration=None,
+        cron_schedule=None,
+        classifiers=None,
+        table_prefix=None,
```
in an operator, it does make sense to have the config as init params.
but here i'd suggest consoldating to the more future-proof and simpler `crawler_kwargs`, a dict param

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,294 @@
+        crawler_run_state = crawler_run['Crawler']['State']
+        return crawler_run_state
+
+    def get_crawler_status(self, crawler_name: str) -> str:
```
`get_crawl_status` -- it's the status of the crawl i.e. the run not the state of the crawler

this would help clarify diff between `get_crawler_status` and `get_crawler_state` which sound like the same thing

you could even do `get_last_crawl_status` to be explicit

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,294 @@
+                    print('Tables Updated: ', metrics['TablesUpdated'])
+                    print('Tables Deleted: ', metrics['TablesDeleted'])
+
+                    return {'Status': crawler_run_status}
```
why not just return `crawler_run_status` instead of wrapping in dict?

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,294 @@
+            self.log.error("Failed to create aws glue crawler, error: %s", general_error)
+            raise
+
+    def initialize_crawler(self):
```
i might consider chopping this method because initialize is a bit confusing and it doesn't really do much.

then in your operator execute you would just call
```python
crawler_name = hook.get_or_create_glue_crawler(**crawler_kwargs)
hook.start_crawler(crawler_name)
hook.await_crawler(crawler_name)
```
which i think is a tad clearer

marshall7m committed:
consolidate params into `config` dict param and add cached conn/hooks

dstandish commented on tests/providers/amazon/aws/hooks/test_glue_crawler.py:
```diff
@@ -0,0 +1,142 @@
+
+
+class TestAwsGlueCrawlerHook(unittest.TestCase):
+    @cached_property
```
isn't setUp a special method that is run at start of each class... i am surprised if this doesn't cause error cus i assume it would try to call `setUp()` but as it is a property this would not work

if you want it to run only once you can implement `setUpClass` instead

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,181 @@
+        :type config = Optional[dict]
+        :return: Name of the crawler
+        """
+        self.get_iam_execution_role(config["Role"])
```
do you need to do anything with this role you are retrieving?  it seems this doesn't have an effect?

marshall7m replied:
It's more of a check to see if the role exists/valid. I was using the glue job hook (`glue.py`) as a reference in which they used that method. Similar to what you said in another suggestion, I think the Glue API exception for invalid crawler IAM role should hopefully be clear enough. I'll check and probably remove this method from the hook.

marshall7m replied:
@dstandish I tested the glue API's create_crawler() method with an invalid role and the exception was misleading because it outputted: ```botocore.errorfactory.InvalidInputException: An error occurred (InvalidInputException) when calling the CreateCrawler operation: Service is unable to assume role arn:aws:iam::111111111111:role/test-foo-role. Please verify role's TrustPolicy```.  The `self.get_iam_execution_role()` outputted a more appropriate exception: ```botocore.errorfactory.NoSuchEntityException: An error occurred (NoSuchEntity) when calling the GetRole operation: The role with name test-foo-role cannot be found.```

dstandish replied:
personally, i would get rid of this method, and just let it fail at get_crawer.  the outcome is the same whether you check role existence first or skip the check --- the job fails, with a message about the role.

including the method suggests that it needs to be there, so it will be unnecessarily confusing.

if you just want to translate an error message, you could catch the error in the appropriate spot and translate it.

but i don't think that's necessary because the error gives enough info.

but that's just me :)

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,181 @@
+                return config["Name"]
+            except Exception as general_error:
+                self.log.error("Failed to create aws glue crawler, error: %s", general_error)
+                raise
```
The nested trys get a bit hard to decipher.  

```suggestion
        crawler_name = config["Name"]
        try:
            self.glue_client.get_crawler(Name=crawler_name)
            self.log.info(f"Crawler '{crawler_name}' already exists; updating crawler.")
            self.glue_client.update_crawler(**config)
        except self.glue_client.exceptions.EntityNotFoundException:
            self.log.info("Creating AWS Glue crawler")
            self.glue_client.create_crawler(**config)
        return crawler_name
```

I think it's better to remove the try except when you are just catching and immediately raising because it impairs readability without benefit --- it will be clear enough in the logs that the operation failed.

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,181 @@
+        try:
+            crawler_run = self.glue_client.start_crawler(Name=crawler_name)
+            return crawler_run
+        except Exception as general_error:
```
again i'm not sure if it's worth catching only to raise...  are the exceptions you get from api not clear enough?

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,181 @@
+            self.log.info("Iam Role Name: %s", role_name)
+            return glue_execution_role
+        except Exception as general_error:
+            self.log.error("Failed to create aws glue crawler, error: %s", general_error)
```
message not correct for this method... but also ... to be a broken record :) ... not sure necessary to catch here

dstandish commented on airflow/providers/amazon/aws/operators/glue_crawler.py:
```diff
@@ -0,0 +1,73 @@
+    :param poll_interval: Time (in seconds) to wait between two consecutive calls to check crawler status
+    :type poll_interval: int
+    :param config = Configurations for the AWS Glue crawler
+    :type config = Optional[dict]
```
not optional

dstandish commented on airflow/providers/amazon/aws/operators/glue_crawler.py:
```diff
@@ -0,0 +1,73 @@
+    :type config = Optional[dict]
+    """
+
+    template_fields = ()
```
if you aren't changing these you can just inherit
but does it make sense to template the `config` param?

dstandish commented on airflow/providers/amazon/aws/operators/glue_crawler.py:
```diff
@@ -0,0 +1,73 @@
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.aws_conn_id = (aws_conn_id,)
```
should be string not tuple

dstandish commented on airflow/providers/amazon/aws/sensors/glue_crawler.py:
```diff
@@ -0,0 +1,54 @@
+        self.crawler_name = crawler_name
+        self.aws_conn_id = aws_conn_id
+        self.success_statuses = ['SUCCEEDED']
+        self.errored_statuses = ['FAILED', 'CANCELLED']
```
these probably make more sense as class constants

dstandish commented on tests/providers/amazon/aws/hooks/test_glue_crawler.py:
```diff
@@ -0,0 +1,142 @@
+        )
+        iam_role = self.hook.get_iam_execution_role(role_name=mock_role_name)
+
+        self.assertIsNotNone(iam_role)
```
perhaps it would be just as easy to verify that the role you mocked is the role retrieved (which is slightly better than just confirming that the value is not None)

dstandish commented on tests/providers/amazon/aws/hooks/test_glue_crawler.py:
```diff
@@ -0,0 +1,142 @@
+    @mock.patch.object(AwsGlueCrawlerHook, "get_iam_execution_role")
+    @mock.patch.object(AwsGlueCrawlerHook, "get_conn")
+    def test_get_or_create_crawler(self, mock_get_conn, mock_get_iam_execution_role):
+        mock_get_iam_execution_role.return_value = mock.MagicMock(Role={'RoleName': mock_role_name})
```
does this have any effect?

dstandish commented on tests/providers/amazon/aws/hooks/test_glue_crawler.py:
```diff
@@ -0,0 +1,142 @@
+    def test_get_or_create_crawler(self, mock_get_conn, mock_get_iam_execution_role):
+        mock_get_iam_execution_role.return_value = mock.MagicMock(Role={'RoleName': mock_role_name})
+
+        mock_glue_crawler = mock_get_conn.return_value.get_crawler()['Crawler']['Name']
```
i think what you are trying to do is mock get_crawler to return  some value, and then confirm that get_or_create returns this same value.  but here you are not correctly mocking a return value -- you are just setting `mock_glue_crawler` equal to a mock object `<MagicMock name='get_conn().get_crawler()[41 chars]936'>`

dstandish commented on tests/providers/amazon/aws/sensors/test_glue_crawler.py:
```diff
@@ -0,0 +1,60 @@
+
+    @mock.patch.object(AwsGlueCrawlerHook, 'get_conn')
+    @mock.patch.object(AwsGlueCrawlerHook, 'get_crawler_state')
+    def test_poke(self, mock_get_crawler_state, mock_conn):
```
you could consider combining these, parameterizing for the various statuses and their expected outcome

marshall7m committed:
1. rm unnecessary try statements from hook
2. rm template fields from operator
3. rm unnecessary try statements from sensor
4. rm get iam role from test hook
5. create crawler statutes tests in test sensor

marshall7m committed:
1. remove unused vars/import
2. change crawler_run_state -> crawler_state (consistent with get_crawler_state)
3. remove crawler_name template field

marshall7m committed:
- Remove empty newline after function doc strings
- Remove unused airflow.configuration import
- Log metrics instead of print metrics for wait_for_crawler_completion()

marshall7m committed:
Silence unnecessary pylint error:
```
Exception arguments suggest string formatting might be intended (raising-format-tuple)
```
in crawler hook AirflowException

mschmo commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,159 @@
+    """
+
+    def __init__(self, poll_interval, *args, **kwargs):
+        self.poll_interval = poll_interval
```
IMO `poll_interval` shouldn't be a required argument for the hook. It makes sense as an argument for the sensor, but the hook I think should be a bit more vague in how it can interact with the Glue API. For instance `poll_interval` is irrelevant if I want to initialize a hook solely to get or create a crawler, and not poll for its status.

I believe it may be better to pass `poll_interval` in as an argument to the `wait_for_crawler_completion` method, which is the only place it's used atm.

mschmo commented on airflow/providers/amazon/aws/sensors/glue_crawler.py:
```diff
@@ -0,0 +1,51 @@
+        self.errored_statuses = ('FAILED', 'CANCELLED')
+
+    def poke(self, context):
+        hook = AwsGlueCrawlerHook(aws_conn_id=self.aws_conn_id, poll_interval=5)
```
More reason IMO to remove `poll_interval` from the hook initialization. This is somewhat confusing because it makes it seem like we're hard-coding a 5 second poll interval, but really we're using the base sensor's `poll_interval` attribute between polls.

mschmo commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,159 @@
+                    print('Estimated Time Left (seconds): ', time_left)
+                    self.poll_interval = time_left
+                else:
+                    print('Crawler should finish soon')
```
Should be using `self.log` instead of `print`.

mschmo commented on airflow/providers/amazon/aws/sensors/glue_crawler.py:
```diff
@@ -0,0 +1,51 @@
+
+class AwsGlueCrawlerSensor(BaseSensorOperator):
+    """
+    Waits for an AWS Glue crawler to reach any of the statuses below
+    'FAILED', 'CANCELLED', 'SUCCEEDED'
```
FYI after the initial run the crawler's last crawl status will **always** be one of these statuses. Do you think it would make more sense to first check that the crawler is in a "READY" state? I feel like this sensor would be much more useful if it behaved similar to the hook's `wait_for_crawler_completion` and confirmed that the crawler was not currently running.

mschmo reviewed (changes requested):
Hey @marshall7m just left a few thoughts! Thanks for taking the time to work on this!

mschmo commented on airflow/providers/amazon/aws/operators/glue_crawler.py:
```diff
@@ -0,0 +1,71 @@
+    @cached_property
+    def hook(self) -> AwsGlueCrawlerHook:
+        """Create and return an AwsGlueCrawlerHook."""
+        return AwsGlueCrawlerHook(self.aws_conn_id)
```
And remember that `AwsGlueCrawlerHook`'s first argument atm is `poll_interval` - so this is actually a bug. But again, I think `poll_interval` should be removed from the `__init__` all together.

marshall7m committed:
- remove `poll_interval` param from crawler hook __init__()
- add `poll_interval` param to crawler hook wait_for_crawler_completion()
- return None for crawler hook check_iam_role() instead of unused iam role
- fix type annotations

marshall7m committed:
fix class function syntax and remove unused json import

marshall7m committed:
- remove unused mock_iam import
- fix empty return statement in crawler sensor

mschmo commented on airflow/providers/amazon/aws/operators/glue_crawler.py:
```diff
@@ -0,0 +1,71 @@
+        self.log.info("Triggering AWS Glue Crawler")
+        self.hook.start_crawler(crawler_name)
+        self.log.info("Waiting for AWS Glue Crawler")
+        self.hook.wait_for_crawler_completion(crawler_name)
```
I think we should pass `self.poll_interval` into this method call otherwise that instance attribute is useless.

marshall7m replied:
Thanks! Currently working on it,  should have a revision up in ~30 mins.

mschmo reviewed (changes requested):
Looking good man! One more comment.

marshall7m committed:
- Add `poll_interval` param to crawler operator wait_for_crawler_completion()
- replace config dict param for get_or_create_crawler with kwargs param: `crawler_kwargs`

mschmo commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,169 @@
+        try:
+            glue_response = self.glue_client.get_crawler(Name=crawler_name)
+            self.log.info("Crawler %s already exists; updating crawler", crawler_name)
+            self.glue_client.update_crawler(**crawler_kwargs)
```
Nice, I overlooked this update API method when implementing a custom Glue hook/operator in my own project. Good to know about!

mschmo reviewed (approved):
LGTM 👍

feluelle commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,169 @@
+    """
+    Interacts with AWS Glue Crawler
+    :param config = Configurations for the AWS Glue crawler
+    :type config = dict
```
```suggestion
```
?

marshall7m replied:
silly me must have somehow mixed up the operator with the hook param

feluelle commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,169 @@
+        Is needed because the current Boto3 (<=1.16.46)
+        glue client create_crawler() method misleadingly catches
+        non-existing role as a role trust policy error.
+        :param role_name = IAM role name
```
```suggestion
        non-existing role as a role trust policy error.

        :param role_name = IAM role name
```
Please put an empty line between description and params.

feluelle replied:
And check this on the rest of your code, please. Thanks :)

feluelle commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,169 @@
+                return glue_response['Crawler']['Name']
+            except self.glue_client.exceptions.InvalidInputException as general_error:
+                self.check_iam_role(crawler_kwargs['Role'])
+                raise AirflowException(general_error)
```
Can you maybe move this into a separate `create_crawler` function?

I would also move the get-or-create logic on the operator level. So that in the execute you would:
```python
if not self.hook.has_crawler(**self.config):
    self.hook.create_crawler(**self.config)
crawler_name = self.hook.get_crawler(**self.config)
```
where `has_crawler` gets the crawler and handles the exception in case it does not exist i.e. returns False.

WDYT?

marshall7m replied:
@feluelle That's a great idea! Seems like it will remove a future pain point in regards to the current `get_or_create_crawler()` function.  Specifically, in a scenario where the user uses the function to just see if the crawler exists but accidentally creates a typo with the crawler_name and ends up creating a new crawler with the same configurations but with a slightly different name.

marshall7m replied:
Maybe the operator should be an if/else since the `get_crawler()` function will also have the Glue API's `update_crawler()` function in it and it would be unnecessary to call `update_crawler()` after running `create_crawler()`

For example:
```
if not self.hook.has_crawler(**self.config):
    crawler_name = self.hook.create_crawler(**self.config)
else:
    crawler_name = self.hook.get_crawler(**self.config)
```

Update:

Or create separate functions for `update_crawler()` and `get_crawler()` within the hook and call them separately in the operator:

```
if not self.hook.has_crawler(**self.config):
    crawler_name = self.hook.create_crawler(**self.config)
else:
    crawler_name = self.hook.get_crawler(**self.config)
    self.hook.update_crawler(**self.config)
```

feluelle replied:
Sure. Sounds good. :)

feluelle replied:
Hmm actually do wee need to call `get_crawler` again if we know that it has the crawler? Can we not just take `crawler_name` (provided) or is this different from the returned one of `get_crawler`?

marshall7m replied:
Good point, no we don't. They are both the same. I'm thinking of deleting `get_crawler()` function from the hook entirely as it's not needed.

marshall7m committed:
- Fix docstrings format and invalid param values
- split hook get_or_create_crawler() into smaller funct
- reflect small hook funcs in operator and test hook
- add sensor get_hook() func to rm repeated calls to crawler hook

Commit:
Fix tests

marshall7m committed:
rm unused imports and add cls decorator

marshall7m committed:
-rm get_crawler() hook func
-add update_crawler() hook func
-add update_crawler() tests

feluelle reviewed (approved):
LGTM 👍

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,216 @@
+        glue client create_crawler() method misleadingly catches
+        a non-existing role as a role trust policy error.
+
+        :param role_name = IAM role name
```
```suggestion
        :param role_name: IAM role name
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,216 @@
+        a non-existing role as a role trust policy error.
+
+        :param role_name = IAM role name
+        :type role_name = str
```
```suggestion
        :type role_name: str
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,216 @@
+        Updates crawler configurations
+
+        :param crawler_kwargs = Keyword args that define the configurations used for the crawler
+        :type crawler_kwargs = any
```
```suggestion
        :param crawler_kwargs: Keyword args that define the configurations used for the crawler
        :type crawler_kwargs: any
```

mik-laj commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,216 @@
+        :return: State of the Glue crawler
+        """
+        crawler = self.glue_client.get_crawler(Name=crawler_name)
+        crawler_state = crawler['Crawler']['State']
```
Are we sure we need separate methods for one field? Can't we generalize it more?  During one iteration of the loop within the `wait_for_crawler_completion ` function, the crawler is fetched 3 times. It's a bit of a lot.

marshall7m replied:
maybe remove `get_crawler_status()` and `get_crawler_state()` and add a `get_crawler()` function that returns the results of the `glue_client.get_crawler()`?  Then we can just index into the response to get the state/status within `wait_for_crawler_completion()`?

something like this for `wait_for_crawler_completion()`:

```
while True:
            crawler = self.get_crawler(crawler_name)
            crawler_state = crawler['Crawler']['State']
            if crawler_state == 'READY':
                self.log.info("State: %s", crawler_state)
                crawler_status = crawler['Crawler']['LastCrawl']['Status']
```

dstandish commented on airflow/providers/amazon/aws/hooks/glue_crawler.py:
```diff
@@ -0,0 +1,216 @@
+            glue_response = self.glue_client.create_crawler(**crawler_kwargs)
+            return glue_response['Crawler']['Name']
+        except self.glue_client.exceptions.InvalidInputException as general_error:
+            self.check_iam_role(crawler_kwargs['Role'])
```
i am not sure why you would want to check iam role if you're gonna raise anyway?
why not just let the exception go uncaught?
otherwise looking good now :)

marshall7m replied:
As mentioned in the `check_iam_role()` docs, the Glue API's `create_crawler()` function doesn't catch non-existing roles properly and throws an inaccurate trust policy error. So the purpose of the`check_iam_role()` is to throw the IAM client's non-existing role error instead of the inaccurate trust policy error. 

But your comment does remind me that I should change the `check_iam_role()` to throw an AirflowException with the non-existing role error so `AirflowException(general_error)` is omitted.

dstandish replied:
i remember this discussion

i would leave it with the original error, which already mentions a problem related to the role when that's the cause of the exception

i think adding this bit of code only adds confusion.  it's kind of a backward side-effecty way of accompishing what you're trying to accomplish.  and the exception won't _always_ be due to role.

if you really must tranlate the error, i would parse the error message for the specific language that appears in this case, and and if you find it, raise an informative message and in other circumstances just raise.  that would be more direct, clear, and explicit.


my 2 cents  🤷‍♂️

marshall7m replied:
Any thoughts on this? @mik-laj @mschmo @feluelle

dstandish replied:
my intention is not to block this pr if no one else cares about this but just want to highlight the issue in case anyone does

feluelle replied:
I would suggest to add @marshall7m comment there, sth. like:

```python
try:
    ...
except self.glue_client.exceptions.InvalidInputException:
    # <add-explanation-here>
    self.check_iam_role(crawler_kwargs['Role'])
    raise
```
I would not raise `AirflowException` tbh. Personally I don't like using this type of exception if it actually is unrelated to airflow. Just re-raise the error in case it has nothing to do with the iam role. If it's because of the iam role it `check_iam_role` will raise a more accurate error.

WDYT?

dstandish replied:
my point is, why create a method `check_iam_role` that serves no purpose other than throwing an error, when the only reason you're calling it is because you _already know_ the role is bad.

in testing we saw that when InvalidInputException is raised due to role, it mentions the role in the error.  so it seems pointless to me.

but, it's also not really going to kill me if this goes in so i'll quit yapping about it after this :)

feluelle replied:
> in testing we saw that when InvalidInputException is raised due to role, it mentions the role in the error. so it seems pointless to me.

Got it. If that's really the case then we do not need and probably should not check the iam role again. We save a request :) Only if it does not mention the role it makes sense.

(sry Daniel I missed the paragraph 😬 )
> i would leave it with the original error, which already mentions a problem related to the role when that's the cause of the exception

marshall7m replied:
After thinking about it, removing the `check_iam_role` method is the right move. As you said @dstandish, the InvalidInputException points the error to the role so that exception should be enough. If the user tries to debug why they got the trust policy error regarding the role, they would probably check if the role they're checking is the right role :).  I'm glad you brought it up, it was a good learning point for me 👍 .

marshall7m replied:
@feluelle Thanks for the input, it's always nice to minimize the number of requests :) 

I'll push the changes sometime today

marshall7m committed:
- rm check_iam_role() from hook
- replace get_crawler_state() & get_crawler_status() with get_crawler()
- add get_crawler to hook tests

marshall7m committed:
fix doc string style

feluelle reviewed (approved):
Very nice! :) 👍
